app/src/modules/transport/transport.c

```diff
@@ -124,108 +248,288 @@ static void connect_work_fn(struct k_work *work)
-		LOG_ERR("Discarding payload since we are not connected to cloud.");
+		LOG_WRN("Discarding payload since we are not connected to cloud");
 	}
 }
 
-/* Function executed when the module enters the connected state. */
-static void connected_entry(void *o)
+/* Handlers for STATE_CONNECTING */
+
+static void state_connecting_entry(void *o)
 {
 	ARG_UNUSED(o);
 
-	enum cloud_status status = CLOUD_CONNECTED;
-	int err = zbus_chan_pub(&CLOUD_CHAN, &status, K_NO_WAIT);
+	LOG_DBG("%s", __func__);
+
+	k_work_reschedule_for_queue(&transport_queue, &connect_work, K_NO_WAIT);
+}
+
+static void state_connecting_run(void *o)
+{
+	struct s_object *state_object = o;
+
+	LOG_DBG("%s", __func__);
+
+	if (state_object->chan == &PRIV_TRANSPORT_CHAN) {
+		enum priv_transport_evt conn_result = *(enum priv_transport_evt *)state_object->msg;
+
+		switch (conn_result) {
+		case CLOUD_CONN_SUCCES:
+			STATE_SET(STATE_CONNECTED);
+
+			return;
+		case CLOUD_CONN_FAILED:
+			STATE_SET(STATE_CONNECTING);
+
+			return;
+		default:
+			break;
+		}
+	}
+}
+
+static void state_connecting_exit(void *o)
+{
+	ARG_UNUSED(o);
+
+	connect_work_cancel();
+}
+
+/* Handler for STATE_CLOUD_CONNECTED. */
+static void state_connected_entry(void *o)
+{
+	ARG_UNUSED(o);
+
+	LOG_DBG("%s", __func__);
+
+	/* Cancel any ongoing connect work when we enter STATE_CLOUD_CONNECTED */
+	connect_work_cancel();
+}
+
+static void state_connected_run(void *o)
+{
+	ARG_UNUSED(o);
+
+	LOG_DBG("%s", __func__);
+}
 
+static void state_connected_exit(void *o)
+{
+	int err;
+
+	ARG_UNUSED(o);
+
+	LOG_DBG("%s", __func__);
+
+	err = nrf_cloud_coap_disconnect();
+	if (err && (err != -ENOTCONN)) {
+		LOG_ERR("nrf_cloud_coap_disconnect, error: %d", err);
+		SEND_FATAL_ERROR();
+	}
+}
+
+
+/* Handlers for STATE_CONNECTED_READY */
+
+static void state_connected_ready_entry(void *o)
+{
+	int err;
+	enum cloud_status cloud_status = CLOUD_CONNECTED_READY_TO_SEND;
+
+	ARG_UNUSED(o);
+
+	LOG_DBG("%s", __func__);
+
+	err = zbus_chan_pub(&CLOUD_CHAN, &cloud_status, K_NO_WAIT);
 	if (err) {
 		LOG_ERR("zbus_chan_pub, error: %d", err);
 		SEND_FATAL_ERROR();
+
+		return;
 	}
+}
 
-	/* Cancel any ongoing connect work when we enter connected state */
-	k_work_cancel_delayable(&connect_work);
+static void state_connected_ready_run(void *o)
+{
+	struct s_object *state_object = o;
+
+	LOG_DBG("%s", __func__);
+
+	if (state_object->chan == &NETWORK_CHAN) {
+		if (NETWORK_STATUS_MSG(state_object) == NETWORK_DISCONNECTED) {
+			STATE_SET(STATE_CONNECTED_PAUSED);
+
+			return;
+		}
+
+		if (NETWORK_STATUS_MSG(state_object) == NETWORK_CONNECTED) {
+			STATE_EVENT_HANDLED();
+
+			return;
+		}
+	}
+
+	if (state_object->chan == &PAYLOAD_CHAN) {
+		int err;
+		struct payload *payload = (struct payload *)state_object->msg;
+
+		LOG_DBG("Sending payload to cloud: %p, len: %d",
+			payload->string, payload->string_len);
+
+		err = nrf_cloud_coap_bytes_send(payload->string, payload->string_len, false);
```

Comment:
It seems unsafe to tolerate all error codes returned from this API.

Reply:
Yes, agree. Prior to rework it didn't check the return code at all. I couldn't find a good way to distinguish between critical and non-critical errors here. Will take another look